ISPN-13615 Create RAFT abstraction

pruivo opened the pull request:
A very basic RAFT abstraction using jgroups-raft. Most of the features
are missing and they will be added later.

https://issues.redhat.com/browse/ISPN-13615

Review notes: the goal for this PR is the introduction of the APIs. I tried to keep them isolated from the "user space" but RAFT is highly coupled with the `Transport`. Open to suggestions here...

Missing features:

* Persistence

JGroups-raft uses LevelDB for persisted logs. Future work involves moving from LevelDB to RocksDB. Also, I'm thinking about adding a "file-based log" to be used when the RocksDB jar is not available (to avoid adding a new dependency).

* Dynamic membership

For the server, add to the REST endpoint a resource to add and remove RAFT members. Extend this work to CLI and Console.

tristantarrant commented on core/src/main/java/org/infinispan/configuration/global/TransportConfigurationBuilder.java:
```diff
@@ -196,8 +202,9 @@ public ThreadPoolConfigurationBuilder remoteCommandThreadPool() {
    public
    void validate() {
       if(attributes.attribute(CLUSTER_NAME).get() == null){
```
There is an `.isNull()` method on the `Attribute`

tristantarrant commented on core/src/main/java/org/infinispan/remoting/transport/raft/RaftStateMachine.java:
```diff
@@ -0,0 +1,57 @@
+    * @param dataInput The {@link DataInput} with the snapshot.
+    * @throws IOException If an I/O error happens.
+    */
+   void readStateFrom(DataInput dataInput) throws IOException;
```
I was wondering why we were using blocking I/O here, but reading the JavaDoc it looks like the `readStateFrom` and `writeStateTo` are called in very specific scenarios, so I'm ok with that.

pruivo replied:
This is just an abstraction for JGroups `StateMachine` API and it is used for snapshotting the state machine when the log reaches a certain size.

tristantarrant commented on core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsRaftManager.java:
```diff
@@ -0,0 +1,189 @@
+   private final Map<String, JgroupsRaftChannel<? extends RaftStateMachine>> raftStateMachineMap = new ConcurrentHashMap<>(16);
+
+   JGroupsRaftManager(GlobalConfiguration globalConfiguration, JChannel mainChannel) {
+      if (mainChannel.getProtocolStack().findProtocol(FORK.class) == null) {
```
You have a method `findFork()` in `JGroupsTransport`. You could make it static and reuse that.

tristantarrant commented on core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java:
```diff
@@ -525,6 +530,57 @@ protected void initChannel() {
+            if (addr instanceof ExtendedUUID) {
+               byte[] v = ((ExtendedUUID) addr).get(key);
+               if (v == null || !Arrays.equals(v, value)) {
+                  log.raftProtocolUnavailable("non managed JGroups channel does not have 'raft-id' set.");
```
`s/non managed/non-managed/`

tristantarrant commented:
When starting the server with this branch, I get
```
2022-05-25 10:10:57,077 ERROR (main) [org.jgroups.JChannel] JGRP000016: exception in channelConnected callback: java.lang.NullPointerException: Cannot invoke "io.micrometer.core.instrument.MeterRegistry.gauge(io.micrometer.core.instrument.Meter$Id, Object, java.util.function.ToDoubleFunction)" because "registry" is null java.lang.NullPointerException: Cannot invoke "io.micrometer.core.instrument.MeterRegistry.gauge(io.micrometer.core.instrument.Meter$Id, Object, java.util.function.ToDoubleFunction)" because "registry" is null
	at io.micrometer.core.instrument.Gauge$Builder.register(Gauge.java:190)
	at org.infinispan.metrics.impl.MetricsCollector.registerMetrics(MetricsCollector.java:188)
	at org.infinispan.metrics.impl.MetricsCollector.registerMetrics(MetricsCollector.java:145)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.lambda$channelConnected$10(JGroupsTransport.java:1590)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1220)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.channelConnected(JGroupsTransport.java:1577)
	at org.jgroups.JChannel.lambda$notifyChannelConnected$1(JChannel.java:1104)
	at java.base/java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:807)
	at java.base/java.util.concurrent.CopyOnWriteArraySet.forEach(CopyOnWriteArraySet.java:425)
	at org.jgroups.JChannel.notifyListeners(JChannel.java:1118)
	at org.jgroups.JChannel.notifyChannelConnected(JChannel.java:1104)
	at org.jgroups.JChannel.connect(JChannel.java:363)
	at org.jgroups.JChannel.connect(JChannel.java:352)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.startJGroupsChannelIfNeeded(JGroupsTransport.java:610)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.start(JGroupsTransport.java:483)
	at org.infinispan.remoting.transport.jgroups.CorePackageImpl$1.start(CorePackageImpl.java:42)
	at org.infinispan.remoting.transport.jgroups.CorePackageImpl$1.start(CorePackageImpl.java:27)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.invokeStart(BasicComponentRegistryImpl.java:617)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:608)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.access$700(BasicComponentRegistryImpl.java:30)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startDependencies(BasicComponentRegistryImpl.java:635)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:599)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.access$700(BasicComponentRegistryImpl.java:30)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.findFactory(BasicComponentRegistryImpl.java:264)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.getComponent0(BasicComponentRegistryImpl.java:130)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.getComponent(BasicComponentRegistryImpl.java:63)
	at org.infinispan.factories.impl.BasicComponentRegistry.getComponent(BasicComponentRegistry.java:74)
	at org.infinispan.counter.impl.CounterModuleLifecycle.cacheManagerStarting(CounterModuleLifecycle.java:133)
	at org.infinispan.factories.GlobalComponentRegistry.modulesManagerStarting(GlobalComponentRegistry.java:280)
	at org.infinispan.factories.GlobalComponentRegistry.access$000(GlobalComponentRegistry.java:64)
	at org.infinispan.factories.GlobalComponentRegistry$ModuleInitializer.start(GlobalComponentRegistry.java:370)
	at org.infinispan.factories.CorePackageImpl$11.start(CorePackageImpl.java:173)
	at org.infinispan.factories.CorePackageImpl$11.start(CorePackageImpl.java:171)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.invokeStart(BasicComponentRegistryImpl.java:617)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:608)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.access$700(BasicComponentRegistryImpl.java:30)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.factories.GlobalComponentRegistry.preStart(GlobalComponentRegistry.java:263)
	at org.infinispan.factories.AbstractComponentRegistry.start(AbstractComponentRegistry.java:248)
	at org.infinispan.manager.DefaultCacheManager.internalStart(DefaultCacheManager.java:766)
	at org.infinispan.manager.DefaultCacheManager.start(DefaultCacheManager.java:734)
	at org.infinispan.server.SecurityActions.lambda$startCacheManager$1(SecurityActions.java:68)
	at org.infinispan.security.Security.doPrivileged(Security.java:56)
	at org.infinispan.server.SecurityActions.doPrivileged(SecurityActions.java:40)
	at org.infinispan.server.SecurityActions.startCacheManager(SecurityActions.java:71)
	at org.infinispan.server.Server.run(Server.java:401)
	at org.infinispan.server.Bootstrap.runInternal(Bootstrap.java:165)
	at org.infinispan.server.tool.Main.run(Main.java:98)
	at org.infinispan.server.Bootstrap.main(Bootstrap.java:50)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.infinispan.server.loader.Loader.run(Loader.java:106)
	at org.infinispan.server.loader.Loader.main(Loader.java:51)
```

pruivo commented:
Updated
* rebased on top of JGroups 5
* updated jgroups-raft to 1.0.8.Final
* added persistence (needs global state enabled, is that ok?) using FileBasedLog

pruivo commented:
fixed thread leak

tristantarrant commented:
There was an OOME in CI

tristantarrant commented:
Merged, thanks
